pkg/proxy/router.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 - 2023 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"time"
+
+	"github.com/fagongzi/goetty/v2"
+	"github.com/matrixorigin/matrixone/pkg/clusterservice"
+	"github.com/matrixorigin/matrixone/pkg/common/moerr"
+	"github.com/matrixorigin/matrixone/pkg/frontend"
+	"github.com/matrixorigin/matrixone/pkg/pb/metadata"
+)
+
+const (
+	tenantLabelKey = "account"
+)
+
+// Router is an interface to select CN server and connects to it.
+type Router interface {
+	// Select selects the best CN server with the selector.
+	Select(label labelInfo) (*CNServer, error)
+
+	// Connect connects to the CN server and returns the connection.
+	// It should take a handshakeResp as a parameter, which is the auth
+	// request from client including tenant, username, database and others.
+	Connect(c *CNServer, handshakeResp *frontend.Packet, t *tunnel) (ServerConn, []byte, error)
```

Comment:
How about separating `Select` and `Connect` to different interface? Fine-grained interface allows external plugin to override selection while keep the default `Connect` implemenetation.

Reply:
This interface could be modified to work with whiltelist from cloud platform. When this pr is merged, you can change it to implement the whitelist function.